Make assembler script work with GCC 7.5.0

MJoergen opened the pull request:
When running this script using GCC 7.5.0, the compiler gives the warning 

```
gcc: warning: monitor.asm: linker input file unused because linking not done
```

and the script produces no output, causing the build to fail.

This is because the gcc frontend examines the filename "monitor.asm" and interprets it as an object file to be fed straight into linking. Any file name with no recognized suffix is treated this way.

The proposed change forces the gcc frontend to interpret the input file as a C file, thus invoking the pre-processor as expected.

sy2002 commented:
@MJoergen Thank you! Highly appreciated.

sy2002 commented:
@MJoergen P.S. Just in case you are on Vivado instead of ISE, switch to branch "develop", there you'll find a Vivado version

MJoergen commented:
> @MJoergen P.S. Just in case you are on Vivado instead of ISE, switch to branch "develop", there you'll find a Vivado version

Thanks for the tip. I am indeed trying to build using Vivado. I'll try the develop branch.

BTW: I have some ideas to greatly reduce the number of LUTs in the register_file and the video_bram. I'll try them in the develop branch and if they work I'll make a PR there.

sy2002 commented:
@MJoergen Re Vivado: use the folder `hw/xilinx/nexys4ddr/Vivado` 
Just in case you have other hardware: I appreciate your port :-) 

Re optimizing register_file and video_bram: Great - would be super cool! 😃 🚀  Please be sure to run the processor test program to check, if the timing still works after your optimizations: `test_programs/cpu_test.asm`

sy2002 commented:
P.S. And one more hint: In the branch develop (not released on master, yet), there is `doc/README.md` with a wealth of info.

MJoergen commented:
@sy2002 : When creating bitstream from Vivado GUI, I get an error during implementation regarding Multiple Drivers, and it fails to generate bitstream. The issue is the bidirectional data port on the CPU, and the use of tristate. Apparently, my version of Vivado (2019.2) won't allow this. Have you found a way to build a bitstream using Vivado ?

sy2002 commented:
@MJoergen : I am also using Vivado 2019.2 (on Ubuntu inside a Parallels VM on a Mac) and I just tried it again: Works like a charm. I attached my bitstream for the Nexys4DDR board here: [vivado_bitstream_nexys.zip](https://github.com/sy2002/QNICE-FPGA/files/4968040/vivado_bitstream_nexys.zip)

Did you open this file (from the develop branch) in the Vivado IDE: `hw/xilinx/nexys4ddr/Vivado/qnice_nexys.xpr` ?

What I just tried: I checked out the development branch to a new folder on my HD just to be on the safe side (starting fresh), then I opened the above mentioned file in Vivado 2019.2 and generated the bitstream that you can download here.

Would you like to open an Issue and we discuss it there?  And/or we could Skype and do some screensharing to find out the differences in our Vivado 2019.2 setup.

sy2002 commented:
P.S: Your symptoms sound like you experimented with any MMIO device attached to that bus? If a device is not enabled, then the bus connection needs to go "high impendence" `(others => 'Z')` . So for example 
```
if en = '1' and we = '0' then
   XYZ ABC
else
   data <= (others => 'Z');
endif
```

MJoergen commented:
Today I learned something new! Just now I tried cloning a fresh version of the repository (branch develop) and followed your instructions. And it worked! The bitstream is generated, and loads to the board, and starts the monitor as expected.

Then I changed one thing: In the synthesis flow settings, I changed "Flatten hierarchy" from "rebuilt" to "none", and generated a new bitstream. However, now the implementation fails with the error about "multiple drivers". So apparently, changing this one setting inhibits the ability of vivado to generate netlist for the bidirectional data port. That was a new one for me! I usually always set flatten hierarchy to none in order to get a better view in the schematic.

Personally, I never use bidirectional ports within a design (only on top level ports). But that is just a matter of style I guess.

I see in the generated netlist that the vga_screen generates nice BRAMs and very little logic, so no need for optimization there. However, the register_file does indeed generate registers. With a few changes to the HDL I can help vivado infer LUTRAM instead of registers. The timing is the same, because reading from LUTRAM is combinatorial just like registers. So next step is to figure out how to get the UART working so I can download the cpu_test program to the board. When I verify the fix works, I'll make a PR.

sy2002 commented:
re "Flatten hierarchy" from "rebuilt to "non": Wow! This is really interessting news for me, too. Thank you for sharing!
